pdns/recursordist/rec_channel_rec.cc

```diff
@@ -1378,17 +1378,25 @@ void doExitGeneric(bool nicely)
 #if defined(__SANITIZE_THREAD__)
   _exit(0); // regression test check for exit 0
 #endif
-  g_log << Logger::Error << "Exiting on user request" << endl;
-  g_rcc.~RecursorControlChannel();
+  // Not safe from a signal handler!
+  // g_slog->withName("runtime")->info(Logr::Notice, "Exiting on user request")
+  static const string msg("Exiting on user request\n");
+  (void)write(STDERR_FILENO, msg.data(), msg.size());
 
   if (!g_pidfname.empty()) {
     unlink(g_pidfname.c_str()); // we can at least try..
   }
 
   if (nicely) {
     RecursorControlChannel::stop = true;
+    {
+      std::unique_lock lock(g_doneRunning.mutex);
```

Comment:
Am I right that the `nicely` path is never called from a signal handler? Otherwise locking there would be a problem. At this point I'm wondering if it would make sense to split this function into one version that might be called from a signal handler and one that can't?

Reply:
Yes makes sense to avoid doing dangerous stuff in the sig handler.